docs/design/validation.md

```diff
@@ -0,0 +1,2048 @@
+
+##### Resource Level Attribute Path Validation Handling
+
+Rather than adjusting the `schema.Attribute` type for this type of validation, it could be forced to the resource (or data source) level. The [Declaring Multiple Attribute Validation for Resources](#declaring-multiple-attribute-validation-for-resources) proposals presented later are revelant for this section. To prevent proposal duplication, please see that section for more details and associated tradeoffs.
+
+#### Declaring Multiple Attribute Validation for Resources
+
+In the previous framework, the `CustomizeDiff` functionality enabled resource (or data source) level validation as a logical catch-all. These proposals cover the next iteration of that type of functionality.
+
+##### `PlanModifications` for Resources
+
+The [Plan Modifications design documentation](./plan-modifications.md) outlines proposals which broadly replace the previous framework's `CustomizeDiff` functionality. See that documentation for considerations and recommendations there. In this proposal for validation, new functions for validation would be provided within that framework, rather than introducing separate handling.
+
+Implementing against that design could prove complex for the framework as they are intended to serve differing purposes. It could also be confusing for provider developers in the same way that `CustomizeDiff` was confusing where differing logical rules applied to differing attribute value and operation scenarios.
+
+##### `AttributeValidations` for Resources
+
+This introduces a new extension interface type for `ResourceType` and `DataSourceType`. For example:
+
+```go
+type DataSourceTypeWithAttributeValidations interface {
+    DataSourceType
+    AttributeValidations(context.Context) AttributeValidators
+}
+
+type ResourceTypeWithAttributeValidations interface {
+    ResourceType
+    AttributeValidations(context.Context) AttributeValidators
+}
+```
+
+Where `AttributeValidators` is a slice of types to be discussed later.
+
+As an example sketch, provider developers could introduce a function that fulfills the new interface with example helpers such as:
+
+```go
+func (t *customResourceType) AttributeValidations(ctx context.Context) AttributeValidators {
+    return AttributeValidators{
+      ConflictingAttributes(*tftypes.AttributePath, *tftypes.Attribute),
+      ConflictingAttributesWithValues(*tftypes.AttributePath, ValueValidator, *tftypes.AttributePath, ValueValidator),
+      PrerequisiteAttribute(*tftypes.AttributePath, *tftypes.AttributePath),
+      PrerequisiteAttributeWithValue(*tftypes.AttributePath, ValueValidator, *tftypes.AttributePath),
+    }
+}
+```
+
+This setup would allow for the framework to provide flexible resource level validation with a low amount of friction for provider developers. Helper functions would be extensible and make the behaviors clear.
+
+#### Defining Attribute Validation Functions
```

Comment:
What's an attribute validation function, and how is it different from a multiple attribute validation function or a single attribute validation function? I'm struggling to keep track of the distinctions you're introducing here. So far I've got:

* single attribute value validations
  * I _think_ this is called value validation, sometimes?
* multiple attribute value validations
  * I _think_ this is called path validation, sometimes?
* attribute validations

Is there any way to make the distinction between these concepts clearer?